### PR TITLE
Move grpc rate limiter from podresource folder

### PR DESCRIPTION
**Pull request by kl52752**
Rate limiter.go file is a generic file implementing grpc Limiter interface. This file can be reuse by other gRPC API not only by podresource.

Change-Id: I905a46b5b605fbb175eb9ad6c15019ffdc7f2563

#### What type of PR is this?
/kind cleanup

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
Rate Limit interface will be used by [KEP](https://github.com/kubernetes/enhancements/pull/4184), moving this file outside the podresource folder so it can be reused.

### CODE REVIEW

**Comment by ffromani** [comment]
> Rate Limit interface will be used by [KEP](https://github.com/kubernetes/enhancements/pull/4184), moving this file outside the podresource folder so it can be reused.

I'm fine with generalizing, but I want to point out that this rate-limiting was tailor-made for podresources, so it enshrines some key assumptions and implicit values that can make the aforementioned generalization nontrivial.
IOW, I'm open and supportive for this effort, but I'd expect changes (if nothing else for better UX) needed and perhaps even some kubelet tunables.

I'll review the changeset ASAP.

File: pkg/kubelet/apis/grpc/ratelimit.go [hunk 1]
```diff
@@ -59,9 +59,9 @@ func LimiterUnaryServerInterceptor(limiter Limiter) grpc.UnaryServerInterceptor
 	}
 }
 
-func WithRateLimiter(qps, burstTokens int32) grpc.ServerOption {
+func WithRateLimiter(qps, burstTokens int32, serviceName string) grpc.ServerOption {
```
Comment (ffromani) [inline pkg/kubelet/apis/grpc/ratelimit.go:51]: I'd put the `serviceName` argument first

File: pkg/kubelet/apis/grpc/ratelimit.go [hunk 2]
```diff
@@ -59,9 +59,9 @@ func LimiterUnaryServerInterceptor(limiter Limiter) grpc.UnaryServerInterceptor
...
 	qpsVal := gotimerate.Limit(qps)
 	burstVal := int(burstTokens)
-	klog.InfoS("Setting rate limiting for podresources endpoint", "qps", qpsVal, "burstTokens", burstVal)
+	klog.InfoS("Setting rate limiting for endpoint", serviceName, "qps", qpsVal, "burstTokens", burstVal)
```
Comment (ffromani) [inline pkg/kubelet/apis/grpc/ratelimit.go:54]: this is wrong (and will be evident at runtime), `InfoS` expect key/value pairs. I'd change to
```
klog.InfoS("Setting rate limiting for endpoint", "service", serviceName, "qps", qpsVal, "burstTokens", burstVal)
```

File: pkg/kubelet/server/server.go [hunk 3]
```diff
@@ -219,7 +219,7 @@ type PodResourcesProviders struct {
 
 // ListenAndServePodResources initializes a gRPC server to serve the PodResources service
 func ListenAndServePodResources(endpoint string, providers podresources.PodResourcesProviders) {
-	server := grpc.NewServer(podresourcesgrpc.WithRateLimiter(podresourcesgrpc.DefaultQPS, podresourcesgrpc.DefaultBurstTokens))
+	server := grpc.NewServer(apisgrpc.WithRateLimiter(apisgrpc.DefaultQPS, apisgrpc.DefaultBurstTokens, "podresource"))
```
Comment (ffromani) [inline pkg/kubelet/server/server.go]: if we change the order of parameters (I think we should) this will need to be changed

**Review by ffromani — COMMENTED** [review]
/lgtm

I have my reservations about reusing this code easily, but I can't see anything wrong with this change, and the new location IMO is a sidegrade wrt the old location.

**Comment by ffromani** [comment]
/lgtm cancel

there's actually a comment which needs attention, once that is addressed I'd re-add my LGTM.

File: pkg/kubelet/apis/grpc/ratelimit.go (hunk 1 above)
Comment (kl52752) [reply pkg/kubelet/apis/grpc/ratelimit.go]: done

File: pkg/kubelet/apis/grpc/ratelimit.go (hunk 2 above)
Comment (kl52752) [reply pkg/kubelet/apis/grpc/ratelimit.go]: done

File: pkg/kubelet/server/server.go (hunk 3 above)
Comment (kl52752) [reply pkg/kubelet/server/server.go]: done

File: pkg/kubelet/server/server.go [hunk 4]
```diff
@@ -219,7 +219,7 @@ type PodResourcesProviders struct {
 
 // ListenAndServePodResources initializes a gRPC server to serve the PodResources service
 func ListenAndServePodResources(endpoint string, providers podresources.PodResourcesProviders) {
-	server := grpc.NewServer(podresourcesgrpc.WithRateLimiter(podresourcesgrpc.DefaultQPS, podresourcesgrpc.DefaultBurstTokens))
+	server := grpc.NewServer(apisgrpc.WithRateLimiter("podresource", podresources.DefaultQPS, podresources.DefaultBurstTokens))
```
Comment (ffromani) [inline pkg/kubelet/server/server.go:222]: the service name should be "podresources" (plural)

**Review by ffromani — COMMENTED** [review]
one last typo please - it's admittedly minor, but the churn to fix later is, out of experience, actually bigger. Other than that LGTM

File: pkg/kubelet/server/server.go (hunk 4 above)
Comment (kl52752) [reply pkg/kubelet/server/server.go]: done, sorry for that :)

**Review by ffromani — COMMENTED** [review]
/lgtm

thanks for all your updates!

**Comment by ffromani** [comment]
let's make sure existing podresources e2e tests they all pass. Should be no brainer.

**Comment by ffromani** [comment]
failure not relevant for lane `pull-kubernetes-node-e2e-containerd`: E2eNode Suite: [It] [sig-node] OOMKiller [LinuxOnly] [NodeConformance] init container The containers terminated by OOM killer should have the reason set to OOMKilled

**Comment by ffromani** [comment]
failures can be unrelated and are not necessarily blocking for this PR (unless the lane is `Required` ofc). I'll review later.

**Comment by kl52752** [comment]
> failures can be unrelated and are not necessarily blocking for this PR (unless the lane is `Required` ofc). I'll review later.

yes, this is the same failure in [issue](https://github.com/kubernetes/kubernetes/issues/121047)

**Review by mrunalp — APPROVED** [review]

**Pull request merged**